README.md
# api
The OpenPalceDatabase API

# API Documentation
Can be found [here](docs/index.md)

# Installation
This set of instructions is for installing the API only. To install a full OPD instance on one machine, follow the instructions found here.


1. Install [Node.js](http://nodejs.org/) (0.10.+ with npm), [git](http://git-scm.com/), and java 7 (we recommend `openjdk-7-jre`).
    ````bash
    #example for Ubuntu
    sudo add-apt-repository ppa:chris-lea/node.js
    sudo apt-get update
    sudo apt-get install nodejs git-core openjdk-7-jre
    ````

1. Install [elasticsearch](http://www.elasticsearch.org/). You may also want to (optionally) install the [elasticsearch-head Plugin](http://mobz.github.io/elasticsearch-head/) to view and manage indexes using your web browser. Make sure to restart elasticsearch if you install the plugin.
    ````bash
    #example for Ubuntu
    wget <link to elasticsearch >
    sudo dpkg -i <elasticsearch.deb file>
    cd /usr/share/elasticsearch/
    sudo ./bin/plugin -install mobz/elasticsearch-head
    sudo service elasticsearch restart
    ````

1. Clone this repository. The root of the cloned repo will be referred to as `<root_dir>` for the rest of these instructions.
    ````bash
    cd <somewhere>
    git clone https://github.com/openplacedatabase/api.git
    ````

1. [Download](http://www.openplacedatabase.org/download) the latest data snapshot.

1. Set any needed environment variables:
    ````bash
    // Only required if you want to use a google map
    OPD_GOOGLE_API_KEY="<Your google maps key>"
    // TODO document the rest of the env vars. For now look in app.js
    ````

1. Add a credentials file called `<root_dir>/.credentials.json` that has your username and password in it.
    ````json
    {
      "<your username>": {
        "pwd":"<a password you don't use anywhere else>",
        "id":"1",
        "name":"Your full name as it will appear in any edits you make"
      }
    }
    ````

1. Install [GDAL](http://www.gdal.org/)
    ````bash
    sudo apt-get install gdal
    ````

1. run `npm install` while in `<root_dir>`

1. Run the Open Place Database by typing `node <root_dir>/app.js`. It will bind to port 8080 by default.

1. Import the latest snapshot by running `node <root_dir>/utils/import.js <path_to_downloaded_zip>.zip -u username -p password`. Note that by default this will populate the `<root_dir>/data` directory, which can swamp your machine if not careful. To use S3 instead make sure to set your environment variables correctly.

Note that if you do not load data you need to create the index using `curl -XPUT 'localhost:9200/places/' -d '{}'`


docs/changes.md
# Changes Documentation

* [Get changes](#get-v0changesfromto) - `GET /v0/changes?{from,to}`

##GET /v0/changes?{from,to}

Get a list of all places and their geojsons that have changed

###Parameters

* `from` - A timestamp to get changes from (*Default* null). If not set changes from the beginning of time will be returned. Please don't do that.
* `to` - A timestamp to get changes until (*Default* null). If not set will return as many changes as it can.

###Examples

**Get Changes from 1389710140336**

````
GET /v0/changes?from=1389710140336
Accept: application/json
````

````
200 OK

{
  "status": {
    "code": 200,
    "msgs": []
  },
  "data": [
    {
      "timestamp": 1389710140336,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389710167169,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724469377,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724585771,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724624560,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724640538,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    }
  ]
}
````

**Get Changes from 1389710140336 to 1389724585771**

````
GET /v0/changes?from=1389898363&to=1389724585771
Accept: application/json
````

````
200 OK

{
  "status": {
    "code": 200,
    "msgs": []
  },
  "data": [
    {
      "timestamp": 1389710140336,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389710167169,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724469377,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    },
    {
      "timestamp": 1389724585771,
      "id": "790af1cb-7e45-4235-aac0-fabf0233edb9"
    }
  ]
}
````

routes/tools.js
module.exports = function(server, settings) {

  var util = require('util'),
      exec = require('child_process').exec,
      _ = require('underscore')._,
      api = require(__dirname + '/../lib/api.js');

  server.post('/v0/tools/convert', function(req, res, next) {
    var imagePath,
        outputPath,
        widthHeight = '',
        errors = [];

    // Set image and output path
    if(req.files.image && req.files.image.path) {
      imagePath = req.files.image.path;
      outputPath = imagePath + '.tif';
    } else {
      errors.push('image is required to be a file');
    }

    // If both width and height are set, use them
    if(req.query.width && req.query.height) {
      
      if(!_.isFinite(req.query.width) || req.query.width < 1 || req.query.width > 5000) {
        errors.push('width must be a number between 1 (inclusive) and 5000(inclusive)');
      }
      if(!_.isFinite(req.query.height) || req.query.height < 1 || req.query.height > 5000) {
        errors.push('width must be a number between 1 (inclusive) and 5000(inclusive)');
      }

      widthHeight = '-outsize '+req.query.width+' '+req.query.height;
    } else if(req.query.width || req.query.height) {
      errors.push('both width and height must be set ');
    }
    
    if(errors.length > 0) {
      //console.log('we have errors');
      return res.send(api.format_return(400,errors));
    }

    // -expand rgb needed for pngs
    var command = util.format('gdal_translate %s %s -co PHOTOMETRIC=RGB -co PROFILE=BASELINE %s',imagePath, widthHeight, outputPath);
    //console.log(command)

    // Call command using exec
    exec(command, function(error, stdout, stderr) {
      // Clean up original image

      if(error) {
        // Error Out
        res.json(error);
      } else {
        // Set headers and pipe image to output
        var stream = fs.createReadStream(outputPath);

        res.setHeader('content-type', 'image/tiff');
        stream.pipe(res);
      }
    });
  });

  server.post('/v0/tools/warp', function(req, res, next) {

    var imagePath,
        tmpPath,
        outputPath,
        gcpObj,
        gcpString = '',
        memoryLimit = '',
        transformOption = '', // auto -> , p1 -> -order 1, p2 -> -order 2, p3 -> -order 3, tps -> -tps, 
        resampleOption = '-rn', // near -> -rn, bilinear -> -rb, cubic -> -rc, cubicspline -> -rcs
        maskOption = '', // -srcnodata '17 17 17' and use masked file
        errors = [];

    // Set image and output paths
    if(req.files.image && req.files.image.path) {
      imagePath = req.files.image.path;
      tmpPath = imagePath + '.vrt';
      outputPath = imagePath + '.tif';
    } else {
      errors.push('image is required to be a file');
    }

    // Get gcp
    if(req.body.gcp) {
      try {
        gcpObj = JSON.parse(req.body.gcp);
      } catch(e) {
        errors.push('gcp must be a valid json object');
      }

      if(gcpObj) {
        errors = errors.concat(validateGCP(gcpObj));
      }

    } else {
      errors.push('gcp is required');
    }

    if(errors.length > 0) {
      console.log('we have errors');
      return res.send(api.format_return(400,errors));
    }

    // Create gcpString
    for(var x in gcpObj) {
      gcpString += '-gcp '+gcpObj[x].x+' '+gcpObj[x].y+' '+gcpObj[x].lat+' '+gcpObj[x].lon+' ';
    }

    // Define commands
    var translateCommand = util.format("gdal_translate -a_srs WGS84 -of VRT %s %s %s", imagePath, tmpPath, gcpString);
    
    // check for colorinterop=pal ? -disnodata 255 or -dstalpha
    var warpCommand = util.format("gdalwarp %s %s %s -dstalpha %s -s_srs WGS84 %s %s -co TILED=YES -co COMPRESS=JPEG -co JPEG_QUALITY=85",
          memoryLimit,
          transformOption,
          resampleOption,
          maskOption,
          tmpPath,
          outputPath);
    // We may not need this one?
    var addoCommand = util.format("gdaladdo -r average %s 2 4 8 16 32 64", outputPath);

    // Translate image to VRT format
    exec(translateCommand, function(error, stdout, stderr) {
      if(error) {
        // Error Out
        res.json(error);
      } else {

        exec(warpCommand, function(error, stdout, stderr) {
          if(error) {
            // Error Out
            res.json(error);
          } else {
            // Set headers and pipe image to output
            var stream = fs.createReadStream(outputPath);

            res.setHeader('content-type', 'image/tiff');
            stream.pipe(res);
          }
        }); // End warp
      }
    }); // End translate

  });

  function validateGCP(json) {
    var errors = [];
    // json must be an array
    if(!_.isArray(json)) {
      return ['gcp must be an array of gcp objects'];
    }

    // Each element must have the proper values
    for(var x in json) {
      var gcp = json[x];
      if(!_.isObject(gcp)) {
        errors.push('gcp['+x+'] is not an object');
        continue;
      }
      
      if(_.intersection(Object.keys(gcp),['lat','lon','x','y']).length != 4) {
        errors.push('gcp['+x+'] must contain x, y lat, and lon elements');
        continue;
      }
      if(!_.isFinite(gcp.x) || gcp.x < 0) {
        errors.push('gcp['+x+'].x must be a number >= 0');
      }
      if(!_.isFinite(gcp.y) || gcp.y < 0) {
        errors.push('gcp['+x+'].y must be a number >= 0');
      }
      if(!_.isFinite(gcp.lat) || gcp.lat < -90 || gcp.lat > 90) {
        errors.push('gcp['+x+'].lat must be a number between -90 and 90');
      }
      if(!_.isFinite(gcp.lon) || gcp.lon < -180 || gcp.lon > 180) {
        errors.push('gcp['+x+'].lon must be a number between -180 and 180');
      }

    }

    // Require at least 3 control points
    if(json.length < 3) {
      errors.push('You must specify at least 3 control points');
    }

    return errors;
  }

};

app.js
var restify = require('restify');
    server = restify.createServer(),
    fs = require('fs'),
    path = require('path');

var settings = {
  backing: process.env.OPD_BACKING ? process.env.OPD_BACKING : 'fs', // Options are s3 or fs
  backing_fs_directory: process.env.OPD_BACKING_FS_DIR ? process.env.OPD_BACKING : path.join(__dirname,'data'),
  backing_s3_bucket: process.env.OPD_BACKING_S3_BUCKET ? process.env.OPD_BACKING_S3_BUCKET : '',
  backing_s3_prefix: process.env.OPD_BACKING_S3_PREFIX ? process.env.OPD_BACKING_S3_PREFIX : '',
  logging: process.env.OPD_LOGGING ? process.env.OPD_LOGGING : 'fs', // Options are s3 or fs
  logging_fs_file: process.env.OPD_LOGGING_FS_FILE ? process.env.OPD_LOGGING_FS_FILE : path.join(__dirname,'changes.log'),
  logging_s3_bucket: process.env.OPD_LOGGING_S3_BUCKET ? process.env.OPD_LOGGING_S3_BUCKET : '',
  logging_s3_prefix: process.env.OPD_LOGGING_S3_PREFIX ? process.env.OPD_LOGGING_S3_PREFIX : '',
  elasticsearch_host: process.env.OPD_ES_HOST ? process.env.OPD_ES_HOST : 'localhost',
  elasticsearch_port: process.env.OPD_ES_PORT ? process.env.OPD_ES_PORT : 9200,
  elasticsearch_index: process.env.OPD_ES_INDEX ? process.env.OPD_ES_INDEX : 'places',
  credentials_file: process.env.OPD_CRED_FILE ? process.env.OPD_CRED_FILE : path.join(__dirname,'.credentials.json')
};

server.use(restify.queryParser({mapParams: false}));

server.use(restify.bodyParser({ mapParams: false, keepExtensions: true }));

// Require all routes
require("fs").readdirSync(__dirname+"/routes").forEach(function(file) {
  require(__dirname+"/routes/" + file)(server, settings);
});

server.listen(8080, function() {
  console.log('listening at %s', server.url);
});